File: app/src/main/java/com/commit451/viewholderarrayadapter/sample/MainActivity.java
package com.commit451.viewholderarrayadapter.sample;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.widget.AutoCompleteTextView;
import android.widget.Spinner;

import com.commit451.viewholderarrayadapter.TextSingleViewHolderArrayAdapter;
import com.commit451.viewholderarrayadapter.TextViewHolderArrayAdapter;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        ArrayList<Cheese> cheeses = load();

        Spinner spinner = (Spinner) findViewById(R.id.spinner);
        CheeseSpinnerAdapter adapter = new CheeseSpinnerAdapter(this, cheeses);
        spinner.setAdapter(adapter);

        Spinner singleSpinner = (Spinner) findViewById(R.id.single_spinner);
        singleSpinner.setAdapter(new CheeseSingleAdapter(this, cheeses));

        Spinner textSpinner = (Spinner) findViewById(R.id.text_spinner);
        textSpinner.setAdapter(new TextViewHolderArrayAdapter<>(this, R.layout.item_spinner_cheese, R.layout. item_spinner_dropdown_cheese, cheeses));

        Spinner textSingleSpinner = (Spinner) findViewById(R.id.text_single_spinner);
        textSingleSpinner.setAdapter(new TextSingleViewHolderArrayAdapter<>(this, R.layout.item_spinner_cheese, cheeses));

        AutoCompleteTextView autoCompleteTextView = (AutoCompleteTextView) findViewById(R.id.text_autocomplete);
        autoCompleteTextView.setAdapter(new CheeseSingleAdapter(this, cheeses));
    }

    private ArrayList<Cheese> load() {
        ArrayList<Cheese> cheeses = new ArrayList<>();
        for (int i = 0; i < 100; i++) {
            cheeses.add(Cheeses.getRandomCheese());
        }
        return cheeses;
    }
}

File: app/src/main/java/com/commit451/viewholderarrayadapter/sample/MainActivity.kt
package com.commit451.viewholderarrayadapter.sample

import android.os.Bundle
import android.widget.AutoCompleteTextView
import android.widget.Spinner
import androidx.appcompat.app.AppCompatActivity
import com.commit451.viewholderarrayadapter.SingleTextViewHolderArrayAdapter
import com.commit451.viewholderarrayadapter.TextViewHolderArrayAdapter
import com.commit451.viewholderarrayadapter.ViewHolderArrayAdapter
import java.util.*

class MainActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        val cheeses = load()

        val spinner = findViewById<Spinner>(R.id.spinner)
        val adapter = ViewHolderArrayAdapter(
            context = this,
            objects = cheeses,
            onCreateViewHolder = { parent ->
                CheeseViewHolder.inflate(parent)
            },
            onCreateDropDownViewHolder = { parent ->
                CheeseDropDownViewHolder.inflate(parent)
            },
            onBindViewHolder = { viewHolder, _, item ->
                viewHolder.bind(item)
            },
            onBindDropDownViewHolder = { viewHolder, _, item ->
                viewHolder.bind(item)
            }
        )
        spinner.adapter = adapter

        val singleSpinner = findViewById<Spinner>(R.id.single_spinner)
        singleSpinner.adapter = CheeseSingleAdapter(this, cheeses)

        val textSpinner = findViewById<Spinner>(R.id.text_spinner)
        textSpinner.adapter = TextViewHolderArrayAdapter(this, R.layout.item_spinner_cheese, R.layout.item_spinner_dropdown_cheese, cheeses)

        val textSingleSpinner = findViewById<Spinner>(R.id.text_single_spinner)
        textSingleSpinner.adapter = SingleTextViewHolderArrayAdapter(this, R.layout.item_spinner_cheese, cheeses)

        val autoCompleteTextView = findViewById<AutoCompleteTextView>(R.id.text_autocomplete)
        autoCompleteTextView.setAdapter(CheeseSingleAdapter(this, cheeses))
    }

    private fun load(): ArrayList<Cheese> {
        val cheeses = ArrayList<Cheese>()
        for (i in 0..99) {
            cheeses.add(Cheeses.randomCheese)
        }
        return cheeses
    }
}
